core/lib/src/rocket.rs

```diff
@@ -805,12 +825,22 @@ impl Rocket {
         // NB: executor must be passed manually here, see hyperium/hyper#1537
         let server = hyper::Server::builder(incoming)
             .executor(runtime.executor())
-            .serve(service);
+            .serve(service)
+            .with_graceful_shutdown(shutdown_receiver.unit_error().boxed().compat());
 
-        // TODO.async: Use with_graceful_shutdown, and let launch() return a Result<(), Error>
-        runtime.block_on(server).expect("TODO.async handle error");
+        runtime.block_on(server)?;
 
-        unreachable!("the call to `block_on` should block on success")
+        Ok(())
+    }
+
+    /// Gracefully shut down the instance of Rocket. Returns a `Result`,
+    /// indicating whether the shutdown request was successfully processed.
+    #[inline(always)]
+    pub fn shutdown(&mut self) -> Result<(), ()> {
+        // TODO How do we want to handle a possible error here?
```

Comment:
Will need a resolution at some point before 0.5 on what we want to do with a possible error.

Reply:
IIRC Senders and Receivers only error if the other side is gone. The sender(s) theoretically can't be dropped before the receiver, so I think `recv()` should never fail. And the sender(s) should ignore send errors, because that can only mean the server already is/has shut down.

Reply:
If that's the case, should it even return a `Result`, or should I just `.expect(...)` the value?

Reply:
In the latest commits, I have replaced it with `.expect("Error sending shutdown signal")`. If someone runs across this, we'll definitely know it.